service/history/shard/context_impl.go

```diff
@@ -496,14 +524,14 @@ func (s *ContextImpl) UpdateHandoverNamespaces(namespaces []*namespace.Namespace
 	}
 }
 
-func (s *ContextImpl) GetTimerMaxReadLevel(cluster string) time.Time {
+func (s *ContextImpl) GetScheduledTaskMaxReadLevel(cluster string) time.Time {
```

Comment:
I can return task.Key. But pass in category won't help here I think. Currently category is just a number and I can't tell if I need to return the time or the taskID just based on the category.

Another option is to make `category` an interface, something like the following
```
type Category interface {
    ID() int
    Type() CategoryType // ImmediateTask or ScheduledTask
    Name() string // for displaying and metrics
}
```
then we can have only one method `GetQueueMaxReadLevel(category Category, cluster string) tasks.Key`
Also with this method, we can store an int64 instead of tasks.Key (which has both TaskID and FireTime field) in shardInfo and we can interpret this number by finding a Category with a matching ID and get its CategoryType.  

Currently the concept of Immediate v.s. Scheduled only exists here (shard context), other places use `tasks.Key.FireTime.IsZero` to determine if the task is a scheduled task or not, which is kinda implicit. We can explicit introduce this concept (CategoryType) to the codebase and do checks based on that.

What's your opinion? cc @yiminc

Reply:
The proposed Category interface looks good to me.

Reply:
Cool. Actually instead of defining an interface, a struct with un-exported fields and exported getter methods should be enough. 

I will send out another pr to make this change and land it before the persistence layer and shard context change.